src/figuretype/trader.c

```diff
@@ -943,12 +966,33 @@ void figure_trade_ship_action(figure *f)
     f->image_id = image_group(GROUP_FIGURE_SHIP) + dir;
 }
 
+static double trade_policy_factor() {
+    double trade_factor = 1;
+
+    if(city_buildings_has_caravanserai() && building_monument_working(BUILDING_CARAVANSERAI)) {
+        building *b = building_get(city_buildings_get_caravanserai());
+
+        int policy = building_monument_module_type(BUILDING_CARAVANSERAI);
+
+        if (building_caravanserai_enough_foods(b) && policy == TRADE_POLICY_3) {
+            int pct_workers = calc_percentage(b->num_workers, model_get_building(b->type)->laborers);
+            if (pct_workers >= 100) { // full laborers
+                trade_factor = 1 + POLICY_3_BONUS_PERCENT/100; // caravan capacity increase by 50%
+            } else if (pct_workers > 0) {
+                trade_factor = 1 + POLICY_3_BONUS_PERCENT/100/2; // caravan capacity increase by 25%
+            }
+        }
+    }
+    return trade_factor;
+}
+
 int figure_trade_land_trade_units()
 {
+    int unit = 8;
     if (building_monument_working(BUILDING_GRAND_TEMPLE_MERCURY)) {
-        return 12;
+        unit = 12;
     }
-    return 8;
+    return (int)(unit * trade_policy_factor());
```

Comment:
Here's the appropriate place:

`return calc_adjust_with_percentage(unit, trade_policy_factor());`